erts/emulator/beam/erl_alloc_util.c

```diff
@@ -2539,6 +2542,122 @@ mbc_alloc(Allctr_t *allctr, Uint size)
 			   1);
     return BLK2UMEM(blk);
 }
+/* Given a pointer and size, try shrink it to be page-aligned (4096 bytes
+ * typically) and make size to be a multiple of page size too. */
+static void ERTS_INLINE
+slice_align_to_page(ErtsMemSlice *slice) {
+    static size_t PAGE_SIZE = 0;
+    if (PAGE_SIZE == 0) {
+        PAGE_SIZE = (size_t)getpagesize(); /* 4096 but YMMV */
+    }
+    if (slice->sz < PAGE_SIZE) {
+        slice->sz = 0;
+        return; /* no deal */
+    }
+    {
+        const size_t PAGE_SIZE_MASK = PAGE_SIZE - 1;
+        void *newp = slice->p;
+        /* How many bytes p is after the page start */
+        size_t hanging_bytes = PAGE_SIZE_MASK & (size_t) slice->p;
+        if (hanging_bytes > 0) {
+            /* not aligned: move up by page size and cut extra bits */
+            newp = (void *) (((size_t) slice->p + PAGE_SIZE) & ~PAGE_SIZE_MASK);
+        }
+        /* p moved forward by (page size minus hanging bytes) so reduce size by
+         * that and then cut to a multiple of page size */
+        slice->sz -= hanging_bytes ? (PAGE_SIZE - hanging_bytes) : 0;
+        slice->sz &= ~PAGE_SIZE_MASK;
+        slice->p = newp;
+    }
+}
+
+
+/* Construct a slice struct for user memory of a block */
+static void ERTS_INLINE
+slice_new(Allctr_t *allocator, Block_t *block, ErtsMemSlice *out) {
+  out->p = BLK2UMEM(block) + allocator->min_block_size;
+  out->sz = BLK_UMEM_SZ(block) - allocator->min_block_size;
+}
+
+
+static void ERTS_INLINE
+slice_subtract(ErtsMemSlice *a, ErtsMemSlice *b) {
+    /* Check if b overlaps a from the bottom */
+    if (b->p + b->sz >= a->p && b->p <= a->p) {
+        void *a_end = a->p + a->sz;
```

Comment:
This won't compile with MSVC, pointer arithmetic on `void*` is undefined behavior and acting like a `char*` is a GCC extension.

Reply:
I'm fine if `.p` field becomes a `char *`
to whoever will be delivering this.
Nice find, thanks!